File: parser.py
import json
import csv
import requests
import json
import time
import sqlite3
from tabulate import tabulate
import os
import pandas as pd
import threading
from datetime import datetime

global inhib, backgroundFetch, ThreadStop, ClockSignal
inhib = 0
backgroundFetch = 0
ThreadStop = 0
ClockSignal = 2

def clear_database():
    if os.path.exists("teams.db"):
        os.remove("teams.db")


def read_data_from_json(json_file):
    global inhib, backgroundFetch
    # Clear existing database
    clear_database()
    # os.system("clear")
    with open(json_file, "r") as file:
        teams_data = json.load(file)

    # Connect to SQLite database
    conn = sqlite3.connect("teams.db")
    cursor = conn.cursor()

    # Create table if it doesn't exist
    cursor.execute(
        """CREATE TABLE IF NOT EXISTS teams (
                        team_name TEXT,
                        team_email TEXT,
                        team_uni TEXT,
                        team_other_uni TEXT,
                        member_name TEXT,
                        member_email TEXT,
                        member_gender TEXT,
                        member_nic TEXT,
                        member_ieee TEXT,
                        member_pno TEXT,
                        member_food_pref TEXT,
                        member_git TEXT,
                        team_gist TEXT
                    )"""
    )

    for team in teams_data:
        team_name = team["team"]["name"]
        team_email = team["team"]["email"]
        team_uni = team["team"]["uni"]
        team_other_uni = team["team"]["otherUni"]
        team_gist = team["quizData"]

        for member in team["members"]:
            member_name = member["name"]
            member_email = member["email"]
            member_gender = member["gender"]
            member_nic = member["nic"]
            member_ieee = member["iee"]
            member_pno = member["no"]
            member_fp = member["food"]
            member_git = member["git"]

            # Insert data into the database
            cursor.execute(
                """INSERT INTO teams 
            (team_name,
             team_email, 
             team_uni, 
             team_other_uni, 
             member_name, 
             member_email, 
             member_gender,
             member_nic,
             member_ieee,
             member_pno,
             member_food_pref,
             member_git,
             team_gist)
                              VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
                (
                    team_name,
                    team_email,
                    team_uni,
                    team_other_uni,
                    member_name,
                    member_email,
                    member_gender,
                    member_nic,
                    member_ieee,
                    member_pno,
                    member_fp,
                    member_git,
                    team_gist,
                ),
            )

            # Commit the changes
            conn.commit()
    query = """
            SELECT * FROM teams 
            """
    df = pd.read_sql_query(query, conn)
    df.to_excel("teams_data.xlsx", index=False)
    df.to_csv("teams_data.csv", index=False)
    # Query to fetch and count all unique teams
    cursor.execute("SELECT DISTINCT team_name,team_uni FROM teams ")
    teams = cursor.fetchall()

    cursor.execute("SELECT COUNT(team_name) FROM teams ")
    Totalteams = cursor.fetchall()

    # Query to fetch and count all unique universities
    cursor.execute(
        "SELECT team_uni, COUNT(*) as uni_count FROM teams GROUP BY team_uni"
    )
    universities = cursor.fetchall()

    # Display the counts as a table
    if backgroundFetch == 0:
        os.system("clear")
        print("- Fetching Live Data -")
        print("\nTeam Data\n")
        print("-" * 30)
        print(tabulate(teams, headers=["Team Name", "University"]))
        print("\nUniversity Count")
        print("-" * 30)
        print(
            tabulate(
                [(uni[0], int(uni[1]) / 4) for uni in universities],
                headers=["University", "Count"],
            )
        )
        print("-" * 30)
        print("Total Teams :", int(Totalteams[0][0]) / 4)
        current_datetime = datetime.now()
        print("Time:", current_datetime)
        if inhib == 1:
            os.system("./parser Override")
        # Close the connection
        conn.close()


def AutoFetchJson():
    response = requests.get(
        "ADD_GIST_LINK_HERE"
    )
    if response.status_code == 200:
        # Parse JSONL content
        lines = response.text.splitlines()
        data = [json.loads(line) for line in lines]

        # Save data to teams.json
        with open("teams.json", "w") as outfile:
            json.dump(data, outfile, indent=4)

        # print("Data saved to teams.json")
    else:
        print("Failed to fetch data from the URL !")
    read_data_from_json("teams.json")


def display():
    print("=============================================")
    print("DUOTHAN 4.0 JSON DATA PARSER                ")
    print("=============================================")
    print("1) AutoCheck Json data")
    print("2) Fetch Data in Background & Drop a SQL Shell")
    print("3) AutoCheck Json data | Low Power Mode")
    print("=============================================")


def backgroundFetchInit():
    global ThreadStop, ClockSignal
    ClockSignal = 2
    while ThreadStop != 0:
        AutoFetchJson()
        time.sleep(ClockSignal)


if __name__ == "__main__":
    print("Duothan4.0-JsonParser")
    inputComm = ""
    while True:
        display()
        checkIn = int(input("Enter Value :"))
        if checkIn == 1:
            ClockSignal = 2
            inhib = 0
            while True:
                AutoFetchJson()
                time.sleep(ClockSignal)
        elif checkIn == 2:
            conn = sqlite3.connect("teams.db")
            cursor = conn.cursor()
            backgroundFetch = 0
            bt = threading.Thread(target=backgroundFetchInit)
            bt.start()
            print("SQL Ready !")
            while inputComm != "exitThread":
                inputComm = input(">")
                cursor.execute(inputComm)
                rows = cursor.fetchall()
                for row in rows:
                    print(str(row).replace(",", "|"))
            cursor.close()
            conn.close()
            backgroundFetch = 1
            bt.stop()
        elif checkIn == 3:
            inhib = 0
            ClockSignal = 5
            while True:
                AutoFetchJson()
                time.sleep(ClockSignal)

    # print("Data saved successfully!")

File: parser.cpp
#include <iostream>
#include <fstream>
#include "./nlohmann/json.hpp"
#include <string>

using json = nlohmann::json;


int writeCSVIndividualData(std::string filename) {
    std::ifstream file(filename);
    if (!file.is_open()) {
        std::cerr << "Failed to Open File !" << "\n";
        return 1;
    }

    std::ofstream csvFile("TeamData.csv");
    csvFile << "Team,Team Email,University,Other University,Member Name,Member NIC,Member Email,Member Food,Member Gender,Member Git,Member NO,Member IEE,Quiz Answer\n";

    std::string line;
    while (std::getline(file, line)) {
        json jsonData = json::parse(line);

        if (jsonData.contains("team")) {
            json teamInfo = jsonData["team"];
            std::string teamName = teamInfo["name"];

            if (jsonData.contains("members")) {
                json members = jsonData["members"];
                for (const auto& member : members) {
                    std::string email = member["email"];
                    std::string food = member["food"];
                    std::string gender = member["gender"];
                    std::string git = member.contains("git") ? member["git"] : "";
                    std::string name = member["name"];
                    std::string nic = member["nic"];
                    std::string no = member.contains("no") ? member["no"] : "";
                    std::string iee = member.contains("iee") ? member["iee"] : "";
                    
                    csvFile << teamName << "," 
                            << teamInfo["email"] << "," 
                            << teamInfo["uni"] << "," 
                            << teamInfo["otherUni"] << "," 
                            << name << "," 
                            << nic << "," 
                            << email << "," 
                            << food << "," 
                            << gender << "," 
                            << git << "," 
                            << no << "," 
                            << iee << "," 
                            << jsonData["quizData"] << "\n";
                }
            }
        }
    }
    csvFile << "\n";
    file.close();
    csvFile.close();
    //system("clear");
    std::cout << "Successfully Wrote Data to CSV File !" <<"\n\n\n";
}



int ValidateData(std::string filename) {
    int invalidCountTeams = 0;
    int invalidCountMembers = 0;
    std::ifstream file(filename);
    if (!file.is_open()) {
        std::cerr << "Failed to Open File !" << "\n";
        return 1;
    }

    // std::ofstream csvFile(filename + ".csv");
    // csvFile << "Team,Team Email,University,Other University,Member Name,Member NIC,Member Email,Member Food,Member Gender\n";

    std::string line;
    while (std::getline(file, line)) {
        json jsonData = json::parse(line);

        if (jsonData.contains("team")) {
            json teamInfo = jsonData["team"];
            std::string teamName = teamInfo["name"];

            if (jsonData.contains("members")) {
                json members = jsonData["members"];
                for (const auto& member : members) {
                    std::string email = member["email"];
                    std::string food = member["food"];
                    std::string gender = member["gender"];
                    std::string git = member["git"];
                    std::string name = member["name"];
                    std::string nic = member["nic"];
                    std::string no = member["no"];
                    //Validating Data Before Writing to CSV
                    // if (teamName.find(',') != std::string::npos){
                    //     std::cout << "Invalid Character found in Json file !" << "\n";
                    //     std::cout << "Problematic Value :" << teamName << "\n";
                    //     return 1;
                    // }
                    for (json::iterator it = teamInfo.begin(); it != teamInfo.end(); ++it) {
                        //std::cout << it.key() << ": " << it.value() << "\n";
                        std::string value = it.value();
                        if (value.find(',') != std::string::npos){
                            // std::cout << "--------------------------------" << "\n";
                            // std::cout << "Invalid Value on Team Details !" << "\n";
                            // std::cout << "Problematic Value :" << it.value() <<"\n";
                            // std::cout << "--------------------------------" << "\n";
                            // invalidCountTeams++;
                            std::string arrayTest = it.value();
                            for (int i = 0; i < sizeof(arrayTest)/sizeof(arrayTest[0]); i++) {
                                if (arrayTest[i] == ',') {
                                    arrayTest[i] = NULL;
                                }
                            }
                            std::cout << "Fixed String : " << arrayTest << "\n";
                        }
                    }

                    for (json::const_iterator it = member.begin(); it != member.end(); ++it) {
                            std::string arrayTest = it.value();
                            for (int i = 0; i < sizeof(arrayTest)/sizeof(arrayTest[0]); i++) {
                                if (arrayTest[i] == ',') {
                                    arrayTest[i] = NULL;
                                }
                            }
                            std::cout << "Fixed String : " << arrayTest << "\n";
                    }
                    // csvFile << teamName << "," << teamInfo["email"] << "," << teamInfo["uni"] << "," << teamInfo["otherUni"]<< "," << name << "," << nic << "," << email << "," << food << "," << gender << "\n";
                }
            }
        }
        std::cout << "Total Invalid Team Data : "<< invalidCountTeams <<" | Total Invalid Member Data :" << invalidCountMembers << "\n";
    }

    file.close();
    // csvFile.close();
}


int readIndividualData(std::string filename) {
    std::ifstream file(filename);
    if (!file.is_open()) {
        std::cerr << "Failed to Open File !" << "\n";
        return 1;
    }

    std::string content((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
    file.close();

    size_t start = 0;
    while (start < content.size()) {
        size_t end = content.find("\n", start);
        if (end == std::string::npos) {
            end = content.size();
        }

        std::string line = content.substr(start, end - start);
        json jsonData = json::parse(line);

        if (jsonData.contains("team")) {
            json teamInfo = jsonData["team"];
            std::string teamName = teamInfo["name"];
            std::cout << "Team :" << teamName << "\n";

            if (jsonData.contains("members")) {
                json members = jsonData["members"];
                std::cout << "Members:\n";

                for (const auto& member : members) {
                    std::cout << "  Name: " << member["name"] << "\n";
                    std::cout << "  NIC: " << member["nic"] << "\n";
                    std::cout << "  Email: " << member["email"] << "\n";
                    std::cout << "  Food: " << member["food"] << "\n";
                    std::cout << "  Gender: " << member["gender"] << "\n";
                    std::cout << "------------------------\n";
                }
            }
        }

        start = end + 1;
    }

}

int displayJson(std::string filename) {
    std::ifstream file(filename);
    if (!file.is_open()){
        std::cerr << "Failed to Open File !" << "\n";
        return 1;
    }
    
    std::cout << "Displaying Json Data...\n";
    
    std::string line;
    while (std::getline(file, line)) {
        json jsonData = json::parse(line);
        std::cout << jsonData.dump(4) << std::endl;
    }
    
    file.close();
    return 0;
}

int mainDisplay() {
    std::cout << "=============================================" << std::endl;
    std::cout << "DUOTHAN 4.0 JSON DATA PARSER                " << std::endl;
    std::cout << "=============================================" << std::endl;
    std::cout << "1) Show json file" << std::endl;
    std::cout << "2) Show Grouped Data" << std::endl;
    std::cout << "3) Write Grouped Data as CSV" << std::endl;
    std::cout << "4) Check File's Availability" << std::endl;
    std::cout << "5) Validate Data" << std::endl;
    std::cout << "6) Automatically Count & Display Teams" << std::endl;
    std::cout << "7) Exit" << std::endl;
    std::cout << "=============================================" << std::endl;
}

int main(int argc, char* argv[]) {
    std::string jsonFileName;
    if (argc < 2) {
        std::cout << "Enter Json FileName :";
        // std::getline(std::cin, jsonFileName);
        std::cin >> jsonFileName;
        std::ifstream file(jsonFileName);
        if (!file.is_open()){
            std::cerr << "Failed to Open File !" << "\n";
            return 1;
        }
        while (jsonFileName != "Exit" || jsonFileName != "exit") {
            std::cin.clear();
            int choice;
            mainDisplay();
            std::cout << "Enter Your Choice :";
            std::cin >> choice;
            switch (choice) {
                case 1:
                    if (displayJson(jsonFileName) == 0){
                        std::cout << "Process Completed Successfully..." <<"\n";
                    } else {
                        std::cout << "An error occured !" <<"\n";
                    }
                    break;
                case 2:
                    readIndividualData(jsonFileName);
                    break;
                case 3:
                    writeCSVIndividualData(jsonFileName);
                    break;
                case 5:
                    ValidateData(jsonFileName);
                case 6:
                    return 0;
                    break;
                default:
                    std::cout << "Input Invalid !";
            }
            std::cin.clear();
        }
    } else {
        std::cout << "! Parser Automatic Override !\n";
        writeCSVIndividualData("teamsForP.json");
        system("cat TeamData.csv");
    }
}
